File: source-connector/src/main/java/sourceconnector/domain/pipeline/factory/PipelineBuilder.java
package sourceconnector.domain.pipeline.factory;

import sourceconnector.domain.log.factory.LogFactory;
import sourceconnector.domain.pipeline.Pipeline;
import sourceconnector.domain.processor.BaseProcessor;
import sourceconnector.repository.file.FileRepository;

import java.util.List;

public interface PipelineBuilder<T> {
  /**
   * Creates a pipeline that consists of one or more processors.
   *
   * @param fileRepository the repository used to retrieve the file from external storage
   * @param filePath the path of the file to be processed
   * @param logFactory the factory used to create {@code Log} instances for pipeline operations
   * @param processors the list of base processors to apply in order
   * @return a constructed {@link sourceconnector.domain.pipeline.Pipeline} instance
   *
   * @throws IllegalStateException if the file cannot be retrieved
   * @throws IllegalArgumentException if {@code processors} is {@code null} or empty
   */
  Pipeline<T> create(
    FileRepository fileRepository,
    String filePath,
    LogFactory logFactory,
    List<BaseProcessor<T>> processors
  );

  /**
   * Creates a pipeline that performs no processing and simply bypasses the input data.
   *
   * @param fileRepository the repository used to retrieve the file from external storage
   * @param filePath the path of the file to be processed
   * @param logFactory the factory used to create {@code Log} instances for pipeline operations
   * @return a constructed {@link sourceconnector.domain.pipeline.Pipeline} instance that bypasses processing
   *
   * @throws IllegalStateException if the file cannot be retrieved
   */
  Pipeline<T> createWithNoProcessor(
    FileRepository fileRepository,
    String filePath,
    LogFactory logFactory
  );
}

File: source-connector/src/main/kotlin/sourceconnector/domain/pipeline/factory/PipelineBuilder.kt
package sourceconnector.domain.pipeline.factory

import offsetmanager.domain.file.FileKey
import sourceconnector.domain.log.factory.LogFactory
import sourceconnector.domain.pipeline.Pipeline
import sourceconnector.domain.processor.BaseProcessor
import sourceconnector.repository.file.FileRepository

interface PipelineBuilder<T> {
  /**
   * Creates a pipeline that consists of one or more processors.
   *
   * @param fileRepository the repository used to retrieve the file from external storage
   * @param fileKey the path of the file to be processed
   * @param logFactory the factory used to create `Log` instances for pipeline operations
   * @param processors the list of base processors to apply in order
   * @return a constructed [Pipeline] instance
   *
   * @throws IllegalStateException if the file cannot be retrieved
   * @throws IllegalArgumentException if `processors` is `null` or empty
   */
  fun create(
    fileRepository: FileRepository,
    fileKey: FileKey,
    logFactory: LogFactory,
    processors: List<BaseProcessor<T>>
  ): Pipeline<T?>

  /**
   * Creates a pipeline that performs no processing and simply bypasses the input data.
   *
   * @param fileRepository the repository used to retrieve the file from external storage
   * @param fileKey the path of the file to be processed
   * @param logFactory the factory used to create `Log` instances for pipeline operations
   * @return a constructed [Pipeline] instance that bypasses processing
   *
   * @throws IllegalStateException if the file cannot be retrieved
   */
  fun createWithNoProcessor(
    fileRepository: FileRepository,
    fileKey: FileKey,
    logFactory: LogFactory
  ): Pipeline<T?>
}
